### DISCUSSION TITLE
MIMXRT1062 how to remove LPUART or USBSERIAL and setup up to use printf()

### DISCUSSION THREAD

**Discussion by 8888clockradio** [micropython/micropython#9671] [Hardware & Peripherals]
Anyone know how to turn off the usb serial in the mimxrt1062 / teensy 4.1 and just tell micropython to use printf() or a way that I can use extern "C" to put in Arduino Serial. commands, I think there's c-library bindings for the commands in teensy 4.1. I have some experimenting to do. But for one, where is the line that I'd comment out to just disable the usb serial, and then also how can I tap into an Arduino external Serial C++ object in extern 'C'. I'm porting the teensy 4.1 library and some Arduino libraries to including the audio library into the port. I'm not sure I want to share my code because it's a lot of modified libraries. If anyone is really interested I'm planning on binding Qt touch embedded and music21 into the kernel with micropython to create an environment for a synth kind of like muse.

**Comment by robert-hh** [comment]
All  I/O to USB Serial is done in the file mphalport.c. Transmitting is easy, because that's just the single function mp_hal_stdout_tx_strn(). Receiving is by interrupt, so it's in tud_cdc_rx_cb(). USB is initialized in main.c by a call to tusb_init(). For Audio the mimxrt port has already I2S support.

**Reply by 8888clockradio** [reply]
What about the IRQ's in the .S startup file. do I want to remove those? They dont conflict but I think they may prevent teensy's usb features from loading

**Reply by robert-hh** [reply]
These IRQ end up in the TinyUSB stack, which in turn calls the cb functions in mphalport.c.  If you want to use USB, you have to reroute them. If not, you might make the cb functions just read the data and discard it.

**Reply by 8888clockradio** [reply]
What about this? https://github.com/adafruit/Adafruit_TinyUSB_Arduino

**Reply by 8888clockradio** [reply]
Also where is the sound port info in the i2s mod? Or where for mimxrt

**Reply by robert-hh** [reply]
The file machine_i2s.c contains the set-up of the i.MX RT I2S module and the Python API to send or receive a digital sound stream. The codec to the analog signal has to be added. There are Python libraries for WM8960 and SGTL5000. For other simple codecs there is not additional control required.
